File: tree/binary_search_tree/0700_search_in_bst.js
/*
You are given the root of a binary search tree (BST) and an integer val.

Find the node in the BST that the node's value equals val and return the 
subtree rooted with that node. If such a node does not exist, return null.
*/

// recursive solution
// Time complexity: O(logN) - average case, O(N) - worst case
// Space complexity: O(logN) - average case, O(N) - worst case
var searchBST = function (root, val) {
  if (!root) return null;
  if (root.val === val) return root;
  else if (root.val > val) return searchBST(root.left, val);
  else return searchBST(root.right, val);
};

File: tree/binary_search_tree/0700_search_in_bst.py
# You are given the root of a binary search tree (BST) and an integer val.

# Find the node in the BST that the node's value equals val and return the subtree rooted with that node. If such a node does not exist, return null.

# https://leetcode.com/problems/search-in-a-binary-search-tree/description/

class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return None
        if root.val == val:
            return root
        elif root.val > val:
            return self.searchBST(root.left, val) 
        elif root.val < val:
            return self.searchBST(root.right, val)